### ISSUE TITLE
[dialog] a11y review

### ISSUE THREAD

**Issue by Westbrook** [adobe/spectrum-web-components#1283]
### Feedback

❌ - focus not moved to dialog when opened
❌ - focus not trapped in the dialog once entered - can still move to base page
❌ Dialog should have role=dialog
❌ The rest of the page should have aria-hidden=true so it is not presented to screen readers - will need to move the dialog to be a sibling of page content
❌ Dialog should have aria-labelledby pointing to the dialog title
❌ Dialog contents may or may not be pointed to by aria-describedby on the dialog - this depends on the content of the dialog and where focus ends up.
❌ = focus does not move to the dialog and no indication that it is a dialog is read

### Path forward

1. Review why focus isn't being moved into overlay. Had been doing that recently.
2. Focus is now trapped in the overlay. Previously was not leveraging suggested API in docs site
3. Add `role='dialog'`. This should only be associated with overlay content that capture the tab order, correct?
4. This has been added since the review
5. Associate title via `aria-labelledby` 
6. What sort of heuristic can be used to decide whether content should be associated via `aria-describedby`? 
7. Solved by no. 1 above.

cc @jnurthen

**Comment by Westbrook** [comment]
@jnurthen can you touch on the following a little more:
> What sort of heuristic can be used to decide whether content should be associated via aria-describedby?

**Comment by jnurthen** [comment]
@Westbrook generally around length and structure of the contents and where the first focusable element is in the dialog. 

You only really have to worry about it if the first (or default) focusable item in the dialog is after the main contents of the dialog.

If the dialog content is short enough to be consumed easily (maybe a couple of sentences) then I would always include it with aria-describedby. If it too long (subjective I know) OR includes any important complex structures (which would get lost when stringified) then you wouldn't want to add aria-describedby.

If you are not adding aria-describedby you would mitigate the issue by either focusing the dialog itself or something in the dialog at the start of the contents so the user can advance through the dialog as desired and not focus any elements following the main contents by default.

**Comment by Westbrook** [comment]
@jnurthen finally looking at this right night. For `aria-label`/`aria-labelledby` or `aria-describedby` to apply, the `role="dialog"` element will need `tabindex`. Is there any time when those attributes are available that the first tab stop of the dialog _should_ be the dialog itself? This should be visible via: https://a11y-dialog--spectrum-web-components.netlify.app/storybook/iframe.html?id=overlay--modal-managed&viewMode=story Here I'm using a heuristic of one element or less (which includes any number of text nodes) are applies at `aria-describedby`.

If we were to place focus at the start of the contents, would you still expect the dialog itself to possess the `aria-label`/`aria-labelledby` and `aria-describedby` attributes?  Would this remove the need to "decide" what content deserves to "describe" the dialog?

**Comment by Westbrook** [comment]
@jnurthen thinking some more about this `aria-describedby` concept. If the focus alway throws to the `role="dialog"` element (baring the use of [autofocus](https://a11y-dialog--spectrum-web-components.netlify.app/storybook/?path=/story/dialog-wrapped--form)), meaning you can easily walk through the various pieces of content, is there benefit to working out a heuristic by which sometimes you add `aria-describedby` and sometimes you don't? What I'm looking at right now would work _roughly_ as follows:

```
<sp-dialog>
    <p>Content up to three</p>
    <p>children will automatically be bound</p>
    <p>to the `aria-describedby` attribute</p>
</sp-dialog>

// becomes

<sp-dialog role="dialog" tabindex=0 aria-describedby="dynamicid-0 dynamicid-1 dynamicid-2">
    <p id="dynamicid-0">Content up to three</p>
    <p id="dynamicid-1">children will automatically be bound</p>
    <p id="dynamicid-2">to the `aria-describedby` attribute</p>
</sp-dialog>
```

```
<sp-dialog>
    <p>Content with more than three</p>
    <p>children will NOT be</p>
    <p>bound to the</p>
    <p>`aria-describedby` attribute</p>
</sp-dialog>

// becomes
<sp-dialog role="dialog" tabindex=0>
    <p>Content with more than three</p>
    <p>children will NOT be</p>
    <p>bound to the</p>
    <p>`aria-describedby` attribute</p>
</sp-dialog>
```

Things get a little more tricky when the "content" is a text node:

```
<sp-dialog>
    This content is in a text node. Should the dialog be bound to itself?
</sp-dialog>

// becomes

<sp-dialog role="dialog" tabindex=0 id="dynamicid-0" aria-describedby="dynamicid-0">
    This content is in a text node. Should the dialog be bound to itself?
</sp-dialog>
```

This likely looses its viability as further content is applied:

```
<sp-dialog>
    <h1 slot="headline">Title</h1>
    This content is in a text node. Should the dialog be bound to itself?
    <sp-button slot="button" variant="secondary">Cancel</sp-button>
    <sp-button slot="button" variant="primary">Confirm</sp-button>
</sp-dialog>

// becomes

<sp-dialog role="dialog" tabindex=0 id="dynamicid-0" aria-describedby="dynamicid-0" aria-labelledby="dynamicid-headline-0">
    <h1 slot="headline" id="dynamicid-headline-0">Title</h1>
    This content is in a text node. Should the dialog be bound to itself?
    <sp-button slot="button" variant="secondary">Cancel</sp-button>
    <sp-button slot="button" variant="primary">Confirm</sp-button>
</sp-dialog>
```
Where not the description is A LOT more thing than I think it should be, but maybe that's expected/intended.

**Comment by jnurthen** [comment]
Wow - very detailed analysis.
I think you've convinced me that adding `aria-describedby` automatically when `autofocus` is not being used is not necessary and indeed could make the experience worse. I'm ok in these cases to focus the `dialog` itself.

Do we then simply require the developer to specify `aria-describedby` on the dialog in the cases where `autofocus` is being used - and where there is additional content in the dialog between the dialog "label" and the autofocused element?

**Comment by Westbrook** [comment]
I'm trying to get a better understanding of what the `aria-describedby` gives a user to help lend a suggestion for that situation. Is the idea that with `aria-describedby` that when the screen reader came to the end of the group that they'd get the description that might lead them to "go back"? I've not been seeing this "group"-centric content as much in NVDA as I do in Voice Over, is there a special way to make that surface? I'll spend some more time with that next week.

I think outlining something like the following would make sense:
> While it is not recommended, if you _do_ need to leverage an `[autocomplete]` element within your dialog, be sure to give your root an `aria-describedby` attribute that points to the primary text content of the dialog.

However, I worry that this bit of accessibility advice is the sort that just get "left off" for lack of clear applicable substance. Not just that I haven't used it in the past myself, but more because any suggestion that can't be programmatically applied runs very close to "no aria is better than bad aria".

Thanks in advance for helping me talk this through!
